File: app/src/main/java/com/example/data/auth/AuthRepository.kt
package com.example.data.auth

import android.content.Context
import com.example.domain.model.AuthProviderType
import com.example.domain.model.AuthUser
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class AuthRepository(private val context: Context) {

    private val _currentUser = MutableStateFlow<AuthUser?>(null)
    val currentUser: StateFlow<AuthUser?> = _currentUser.asStateFlow()

    fun signInAsGuest(): Result<AuthUser> {
        val user = _currentUser.value ?: createLocalGuestUser()
        _currentUser.value = user
        return Result.success(user)
    }

    private fun createLocalGuestUser(): AuthUser {
        val randomHex = (1000..9999).random()
        return AuthUser(
            uid = "guest_$randomHex",
            email = null,
            displayName = "Guest Analyst #$randomHex",
            photoUrl = null,
            isAnonymous = true,
            provider = AuthProviderType.GUEST
        )
    }

    fun signOut() {
        _currentUser.value = null
    }

    companion object {
        @Volatile
        private var instance: AuthRepository? = null

        fun getInstance(context: Context): AuthRepository {
            return instance ?: synchronized(this) {
                instance ?: AuthRepository(context.applicationContext).also { instance = it }
            }
        }
    }
}

File: web-app/src/data/auth/AuthRepository.ts
import { AuthProviderType, AuthUser } from '../../domain/models';

const AUTH_STORAGE_KEY = 'aegis_auth_user';

export class AuthRepository {
  private static instance: AuthRepository;
  private user: AuthUser | null = null;
  private listeners: ((user: AuthUser | null) => void)[] = [];

  private constructor() {
    this.loadSession();
  }

  public static getInstance(): AuthRepository {
    if (!AuthRepository.instance) {
      AuthRepository.instance = new AuthRepository();
    }
    return AuthRepository.instance;
  }

  private loadSession() {
    try {
      const stored = localStorage.getItem(AUTH_STORAGE_KEY);
      if (stored) {
        this.user = JSON.parse(stored);
      }
    } catch {
      this.user = null;
    }
  }

  public getCurrentUser(): AuthUser | null {
    return this.user;
  }

  public subscribe(listener: (user: AuthUser | null) => void): () => void {
    this.listeners.push(listener);
    listener(this.user);
    return () => {
      this.listeners = this.listeners.filter(l => l !== listener);
    };
  }

  private notify() {
    for (const l of this.listeners) {
      l(this.user);
    }
  }

  public async signInAsGuest(): Promise<AuthUser> {
    const guestUser: AuthUser = {
      uid: `guest_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`,
      displayName: 'Authenticated Analyst',
      email: null,
      isAnonymous: true,
      provider: AuthProviderType.GUEST,
      createdAt: Date.now()
    };

    this.user = guestUser;
    try {
      localStorage.setItem(AUTH_STORAGE_KEY, JSON.stringify(guestUser));
    } catch {}
    this.notify();
    return guestUser;
  }

  public signOut() {
    this.user = null;
    try {
      localStorage.removeItem(AUTH_STORAGE_KEY);
    } catch {}
    this.notify();
  }
}
